libbpf-tools/xfsslower.bpf.c

```diff
@@ -0,0 +1,159 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020 Wenbo Zhang
+#define BPF_NO_PRESERVE_ACCESS_INDEX
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include "xfsslower.h"
+
+#define NULL    0
+
+const volatile pid_t targ_tgid = 0;
+const volatile __u64 min_lat = 0;
+
+struct piddata {
+	u64 ts;
+	loff_t start;
+	loff_t end;
+	struct file *fp;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, 8192);
+	__type(key, u32);
+	__type(value, struct piddata);
+} start SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(u32));
+	__uint(value_size, sizeof(u32));
+} events SEC(".maps");
+
+static __always_inline int
+probe_entry(struct file *fp, loff_t s, loff_t e)
+{
+	u64 id = bpf_get_current_pid_tgid();
+	struct piddata piddata = {};
+	u32 tgid = id >> 32;
+	u32 pid = id;
+
+	if (targ_tgid && targ_tgid != tgid)
+		return 0;
+
+	piddata.ts = bpf_ktime_get_ns();
+	piddata.start = s;
+	piddata.end = e;
+	piddata.fp = fp;
+	if (piddata.fp)
+		bpf_map_update_elem(&start, &pid, &piddata, 0);
+	return 0;
+}
+
+SEC("kprobe/xfs_file_read_iter")
+int BPF_KPROBE(kprobe__xfs_file_read_iter, struct kiocb *iocb,
+	       struct iov_iter *to)
+{
+	struct file *fp = BPF_CORE_READ(iocb, ki_filp);
+	loff_t start = BPF_CORE_READ(iocb, ki_pos);
+
+	return 0;
+}
+
+SEC("kprobe/xfs_file_write_iter")
+int BPF_KPROBE(kprobe__xfs_file_write_iter, struct kiocb *iocb,
+	       struct iov_iter *from)
+{
+	struct file *fp = BPF_CORE_READ(iocb, ki_filp);
+	loff_t start = BPF_CORE_READ(iocb, ki_pos);
+
+	return probe_entry(fp, start, 0);
+}
+
+SEC("kprobe/xfs_file_open")
+int BPF_KPROBE(kprobe__xfs_file_open, struct inode *inode, struct file *file)
+{
+	return probe_entry(file, 0, 0);
+}
+
+SEC("kprobe/xfs_file_fsync")
+int BPF_KPROBE(kprobe__xfs_file_fsync, struct file *file, loff_t start,
+	       loff_t end, int fdatasync)
+{
+	return probe_entry(file, start, end);
+}
+
+static __always_inline int probe_exit(struct pt_regs *ctx, char type)
+{
+	u64 id = bpf_get_current_pid_tgid();
+	u64 end_ns = bpf_ktime_get_ns();
+	struct piddata *piddatap;
+	struct event event = {};
+	struct file *fp = NULL;
+	struct qstr qs = {};
```

Comment:
nit: no need for initialization if you are setting all those values unconditionally later

Reply:
@anakryiko Thank you for pointing this, I'll remove those unnecessary init, and I've another two questions, the first is about coding style.

 ```c
u64 id = bpf_get_current_pid_tgid();
u64 end_ns = bpf_ktime_get_ns();
struct piddata *piddatap;
struct event event = {}; // if not init, the verifier will complain, I'll describe it as the second question
struct dentry *dentry;
const u8 *qs_name_ptr;
u32 qs_len;
struct file *fp;
u32 tgid = id >> 32;
u32 pid = id;
u64 delta_us;
```

I heard that the order of variable declarations is preferably an inverted Christmas tree, but some variables may be related to each other but may break it, eg

```c
const u8 *qs_name_ptr;
u32 qs_len;
```

and
```c
u32 tgid = id >> 32;
u32 pid = id;
```

Is it ok here and is there a better order?

The second is if I removed `event`'s initialization. The verifier will complain:

> invalid indirect read from stack off -88+85 size 88

Although all members of `event` have been assigned before passing to  `bpf_perf_event_output`


```c
struct event {
        char file[DNAME_INLINE_LEN];
        char task[TASK_COMM_LEN];
        __u64 delta_us;
        __u64 end_ns;
        __s64 offset;
        ssize_t size;
        pid_t tgid;
        char type;
};

fp = piddatap->fp;
dentry = BPF_CORE_READ(fp, f_path.dentry);
qs_len = BPF_CORE_READ(dentry, d_name.len);
qs_name_ptr = BPF_CORE_READ(dentry, d_name.name);
bpf_probe_read_str(&event.file, sizeof(event.file), qs_name_ptr);  // 1

bpf_get_current_comm(&event.task, sizeof(event.task));  // 2
event.delta_us = delta_us;  // 3
event.end_ns = end_ns;  // 4
event.offset = piddatap->start;  // 5
if (type != TRACE_FSYNC)
        event.size = PT_REGS_RC(ctx);  // 6
else
        event.size = piddatap->end - piddatap->start;  // 6
event.type = type;   // 7
event.tgid = tgid;    // 8
```

Is there a way for verifier to be aware of this?

Reply:
> I heard that the order of variable declarations is preferably an inverted Christmas tree, but some variables may be related to each other but may break it, eg

There is such preference in netdev kernel tree, but that preference is overrated. All other things being equal, reverse Christmas tree is somewhat neater-looking. But if the style gets in a way (e.g., when you need a specific order of initializing variables), then do it the way you need it.

> The second is if I removed `event`'s initialization. The verifier will complain:
> 
> > invalid indirect read from stack off -88+85 size 88
> 

argh... it's a 3 byte padding at the end. I guess that `= {}` initialization is needed here, unfortunately. Either that, or explicit char _padding[3] at the end of your struct, but it will be a bit harder to maintain (because you'd need to re-specify padding every time you add or shuffle around fields). Pick your poison, I guess. No big deal to initialize struct, it's just few extra instructions, but I just thought you wouldn't need it, but seems like you do.

Reply:
Get it, thanks.